crates/uv-resolver/src/universal_marker.rs

```diff
@@ -13,74 +16,217 @@ use uv_pep508::{MarkerEnvironment, MarkerTree};
 ///
 /// A universal marker evaluates to true only when *both* its PEP 508 marker
 /// and its conflict marker evaluate to true.
-#[derive(Debug, Default, Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord)]
+#[derive(Default, Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord)]
 pub struct UniversalMarker {
-    pep508_marker: MarkerTree,
-    conflict_marker: MarkerTree,
+    /// The full combined PEP 508 and "conflict" marker.
+    ///
+    /// In the original design, the PEP 508 marker was kept separate
+    /// from the conflict marker, since the conflict marker is not really
+    /// specified by PEP 508. However, this approach turned out to be
+    /// bunk because the conflict marker vary depending on which part of
+    /// the PEP 508 marker is true. For example, you might have different
+    /// a different conflict marker for one platform versus the other. The
+    /// only way to resolve this is to combine them both into one marker.
+    ///
+    /// The downside of this is that since conflict markers aren't part of
+    /// PEP 508, combining them is pretty weird. We could combine them into
+    /// a new type of marker that isn't PEP 508. But it's not clear what the
+    /// best design for that is, and at the time of writing, it would have
+    /// been a lot of additional work. (Our PEP 508 marker implementation is
+    /// rather sophisticated given its boolean simplification capabilities.
+    /// So leveraging all that work is a huge shortcut.) So to accomplish
+    /// this, we technically preserve PEP 508 compatibility but abuse the
+    /// `extra` attribute to encode conflicts.
+    ///
+    /// So for example, if a particular dependency should only be activated
+    /// on `Darwin` and when the extra `x1` for package `foo` is enabled,
+    /// then its "universal" marker looks like this:
+    ///
+    /// ```text
+    /// sys_platform == 'Darwin' and extra == 'extra-3-foo-x1'
```

Comment:
What is "3" here?

Reply:
I thought I explained this somewhere, but it might have only been in a commit message. So I added a code comment here too:

```rust
    /// Why `extra-3-foo-x1`?
    ///
    /// * The `extra` prefix is there to distinguish it from `group`.
    /// * The `3` is there to indicate the length of the package name,
    /// in bytes. This isn't strictly necessary for encoding, but is required
    /// if we were ever to need to decode a package and extra/group name from a
    /// conflict marker.
    /// * The `foo` package name ensures we namespace the extra/group name,
    /// since multiple packages can have the same extra/group name.
```

Does that help?